Draft - WIP to get weather from geolocation (#3)
ericapisani/redwoodjs-weather-demo: README.md

# README

> 

Reviewer: [pebble] In this first, introductory, section I expected to see information about this project as a whole. What is its goal and how does it get to that goal (what tech/architecture). This will help folks know if they want to dig into the project further or not.

Reviewer 2: I've added an introductory section so marking this as resolved, but if there's something that could be done to make it better let me know and I'll make the changes

Run with `netlify dev` locally instead of `yarn rw dev` to test Netlify Edge Functions.

> Note: If editing code while running `netlify dev`, the API server will try to restart but may get a warning that port is in use. You will need to stop and restart `netlify dev` manually.

Reviewer: [pebble] These sentences are relevant to local dev only. I recommend creating a `Local dev` section somewhere below `Requirements` that include these sentences and maybe the `Setting up a database locally` section.

Structuring the README into local dev actions vs deploy/production actions could make it clearer what level of effort folks need to invest.

## Requirements
* A Netlify account. Sign up for free [here](https://app.netlify.com/signup).
* An OpenWeather API key. Get one [here](https://openweathermap.org/api).
* A Postgres database. You can either set one up locally, or use [Railway](https://railway.app/) to provision a temporary one.

## Dataset

> The example uses the [World Cities Dataset](https://simplemaps.com/data/world-cities) from [SimpleMaps](https://simplemaps.com) which "contains demographic details of about 15,000 cities around the world. The location of the cities, the countries to which the City belongs to, its populations etc."

Reviewer: [sand] It wasn't immediately clear to me what "The example" referred to. I assume it mean the project in this repo? If that's true, something like this may be a bit clearer:
```suggestion
This project uses the [World Cities Dataset](https://simplemaps.com/data/world-cities) from [SimpleMaps](https://simplemaps.com) which "contains demographic details of about 15,000 cities around the world. The location of the cities, the countries to which the City belongs to, its populations etc."
```

Reviewer 2: Yeah that was the intention, good idea to clarify 👍🏻

# README

> Run with `netlify dev` locally instead of `yarn rw dev` to test Netlify Edge Functions.

Reviewer 2: Might be worth having a link to the Edge functions documentation for further reading by whoever is trying out this demo

```suggestion
Run with `netlify dev` locally instead of `yarn rw dev` to test [Netlify Edge Functions](https://docs.netlify.com/netlify-labs/experimental-features/edge-functions/).
```

  }
```

## Database setup

> It's recommended to use [`Postgres`] for your database so the schema can use `BigInt` and `Float` nicely.

Reviewer: [sand] Was this meant to be a link to the Postgres website?
```suggestion
It's recommended to use [`Postgres`](https://www.postgresql.org/) for your database so the schema can use `BigInt` and `Float` nicely.
```

Reviewer 2: I don't think so, but I think it's worth the link to it regardless.

> It's recommended to use [`Postgres`] for your database so the schema can use `BigInt` and `Float` nicely.

Reviewer: [dust] We seem to format it as `Postgres` in some instances and just Postgres in others. Not a big deal, but consistency one way or the other would be good.

Also, the seed script uses Prisma's [`createMany()`](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#createmany) to bulk load records in batches which is significantly faster that loading data row by row -- which is important because we have ~43,000 world cities in the dataset.

> Unfortunately, `createMany` is not supported by SQLite.

Reviewer: [sand] A minor addition to emphasize why Postgres is preferred. Based on the wording, I assume SQLite still works -- just not as well.
```suggestion
Unfortunately, `createMany` is not supported by SQLite. If you use a SQLite database, loading records will take significantly longer.
```

### Setting up a database locally

> 

Reviewer: [question] Are there any "gotchas" in using a local Postgres database that we might want to surface here? (I have no idea, ha)

Reviewer 2: I'm not sure as I haven't dabbled much in databases, but @dthyresson might be aware of some as it pertains to developing with the Redwood framework

Reviewer 2: Chatted with David, no known problems so we're all good here 🚀

#### OSX

> The easiest way to get Postgres up ad running locally on OSX is [Postres.app](https://postgresapp.com).

Reviewer: [dust] Missing `n` and `g` here.
```suggestion
The easiest way to get Postgres up and running locally on OSX is [Postgres.app](https://postgresapp.com).
```

> You can then set you database configuration settings to two different local databases, like:

Reviewer: [question] Will most readers know what to do after this? Is there more they need to connect to the database in the app? Is there a tutorial/docs we could point them to? Right now, it kind of feels like a "draw the rest of the owl" situation, but I don't know much about Postgres.

Reviewer 2: They shouldn't need to do more beyond applying the Prisma schema and running the seed script, but I noticed that those instructions are only mentioned in the 'railway' section.

I haven't pushed these changes yet but I'm creating another section for 'populating the data in the database after set up' which lists the commands

TEST_DATABASE_URL=postgresql://postgres:postgres@localhost:5432/redwoodjs-weather-demo-test
```

### Setting up a database on Railway

If you'd rather not worry about setting up a Postgres database on your computer, you can easily provision a Postgres database on [Railway for free](https://railway.app/)(for 24 hours).

After provisioning a database on Railway, copy the 'database connection URL' that Railway provides. You can then use this as your `DATABASE_URL` value in your local environment.

> To apply the Prisma schema to the provisioned databse, run `yarn rw prisma db push`.

Reviewer: [dust] Missing `a`.
```suggestion
To apply the Prisma schema to the provisioned database, run `yarn rw prisma db push`.
```

}
```

## Running Edge Functions in GraphQL

```
[[edge_functions]]
path = "/.redwood/functions/graphql"
function = "geolocate"
```

## Testing Edge functions

1. Place edge function in `netlify/edge-functions`

> 2. In `netlify.toml` your path is to your serverless function and then declare the edge function to run for that path. Be sure to include the deploy path where functions live, like `/.redwood/functions/` or `/.netlify/functions/`

Reviewer 2: Because I think this has the potential to lead to a little bit of confusion, I think it's worth noting here that the `path` is meant to be the [path on the site](https://docs.netlify.com/netlify-labs/experimental-features/edge-functions/get-started/#create-an-edge-function), not the file path and giving more context on why the lambda function path is needed for this demo

For instance, if the URL was `our-redwoodsite.com/weather` that invokes the `geolocate` normal lambda function, and you also wanted to invoke the `geolocate` edge function as part of handling that request, then the entry below would read:

```
[[edge_functions]]
path = "/weather"
function = "geolocate"
```

```
[[edge_functions]]
path = "/.redwood/functions/geolocate"
function = "geolocate"
```

---

> # README

Reviewer 2: Something that we can clean up later, but we should remove the second 'README' heading

}
```

## Running Edge Functions in GraphQL

Add the following to your `netlify.toml` file:

```
[[edge_functions]]
path = "/.redwood/functions/graphql"
function = "geolocate"
```

> Adding this to the `.toml` file means that when a request is made to the `path`, it will invoke the `geolocate` edge function as part of the request.

Reviewer: [dust] Capitalize Edge Function for consistency.
```suggestion
Adding this to the `.toml` file means that when a request is made to the `path`, it will invoke the `geolocate` Edge Function as part of the request.
```

## Testing Edge Functions

> 1. Place edge function in `netlify/edge-functions`

Reviewer: [dust] Capitalize Edge Function for consistency.
```suggestion
1. Place Edge Function in `netlify/edge-functions`
```

> 2. In `netlify.toml`, your path is to your serverless function. In order to declare that the edge function run for that path, you specify the function name as the `function` value. Be sure to include the deploy path where functions live, like `/.redwood/functions/` or `/.netlify/functions/`

Reviewer: [pebble] Breaking this up a bit so that it's easier to follow.
```suggestion
2. In `netlify.toml`, set the `path` value to your serverless function's path. 
  Be sure to include the deploy path where functions live, like `/.redwood/functions/` or `/.netlify/functions/`.
3. Set the `function` value to the name of your Edge Function.
```

```
[[edge_functions]]
path = "/.redwood/functions/geolocate"
function = "geolocate"
```

> Additional documentation on edge functions can be found [here](https://docs.netlify.com/netlify-labs/experimental-features/edge-functions/get-started/)

Reviewer: [dust] Capitalize Edge Functions for consistency.
```suggestion
Additional documentation on Edge Functions can be found [here](https://docs.netlify.com/netlify-labs/experimental-features/edge-functions/get-started/)
```
